File: Snake/snake.py
import pygame

MOVE_DISTANCE = 20
UP = 90
DOWN = 270
LEFT = 180
RIGHT = 0

class Snake:
    def __init__(self):
        self.segments = []
        self.create_snake()
        self.head = self.segments[0]
        self.heading = RIGHT

    def create_snake(self):
        for position in [(0, 0), (-20, 0), (-40, 0)]:
            self.add_segment(position)

    def add_segment(self, position):
        segment = {
            'x': 300 + position[0],  # Center position
            'y': 300 + position[1],
            'rect': pygame.Rect(300 + position[0], 300 + position[1], 20, 20)
        }
        self.segments.append(segment)

    def extend(self):
        # Fix: Extract x,y coordinates from the last segment
        last_segment = self.segments[-1]
        last_position = (last_segment['x'] - 300, last_segment['y'] - 300)
        self.add_segment(last_position)

    def move(self):
        for seg_num in range(len(self.segments) - 1, 0, -1):
            new_x = self.segments[seg_num - 1]['x']
            new_y = self.segments[seg_num - 1]['y']
            self.segments[seg_num]['x'] = new_x
            self.segments[seg_num]['y'] = new_y
            self.segments[seg_num]['rect'].x = new_x
            self.segments[seg_num]['rect'].y = new_y

        if self.heading == UP:
            self.head['y'] -= MOVE_DISTANCE
        elif self.heading == DOWN:
            self.head['y'] += MOVE_DISTANCE
        elif self.heading == LEFT:
            self.head['x'] -= MOVE_DISTANCE
        elif self.heading == RIGHT:
            self.head['x'] += MOVE_DISTANCE

        self.head['rect'].x = self.head['x']
        self.head['rect'].y = self.head['y']

    def up(self):
        if self.heading != DOWN:
            self.heading = UP

    def down(self):
        if self.heading != UP:
            self.heading = DOWN

    def left(self):
        if self.heading != RIGHT:
            self.heading = LEFT

    def right(self):
        if self.heading != LEFT:
            self.heading = RIGHT

    def check_food_collision(self, food):
        return self.head['rect'].colliderect(food.rect)

    def check_wall_collision(self, screen_width, screen_height):
        return (self.head['x'] < 0 or self.head['x'] >= screen_width or
                self.head['y'] < 0 or self.head['y'] >= screen_height)

    def check_self_collision(self):
        for segment in self.segments[1:]:
            if self.head['rect'].colliderect(segment['rect']):
                return True
        return False

    def reset(self):
        self.segments.clear()
        self.create_snake()
        self.head = self.segments[0]
        self.heading = RIGHT

    def draw(self, screen):
        for segment in self.segments:
            pygame.draw.rect(screen, (255, 255, 255), segment['rect'])

File: JSnake/Snake.js
// Snake movement constants
export const MOVE_DISTANCE = 20;
export const UP = 'UP';
export const DOWN = 'DOWN';
export const LEFT = 'LEFT';
export const RIGHT = 'RIGHT';

export class Snake {
  constructor() {
    this.cellSize = 20;
    this.segments = [];
    this.collisionDetected = false;
    this.heading = RIGHT;
    this.createSnake();
    this.head = this.segments[0];
  }

  createSnake() {
    const positions = [[0, 0], [-20, 0], [-40, 0]];
    positions.forEach(position => this.addSegment(position));
  }

  addSegment(position) {
    const segment = {
      x: 200 + position[0], // Center position
      y: 200 + position[1],
      width: this.cellSize,
      height: this.cellSize
    };
    this.segments.push(segment);
  }

  extend() {
    const lastSegment = this.segments[this.segments.length - 1];
    const lastPosition = [lastSegment.x - 300, lastSegment.y - 300];
    this.addSegment(lastPosition);
  }

  move() {
    for (let segNum = this.segments.length - 1; segNum > 0; segNum--) {
      const newX = this.segments[segNum - 1].x;
      const newY = this.segments[segNum - 1].y;
      this.segments[segNum].x = newX;
      this.segments[segNum].y = newY;
    }

    if (this.heading === UP) {
      this.head.y -= MOVE_DISTANCE;
    } else if (this.heading === DOWN) {
      this.head.y += MOVE_DISTANCE;
    } else if (this.heading === LEFT) {
      this.head.x -= MOVE_DISTANCE;
    } else if (this.heading === RIGHT) {
      this.head.x += MOVE_DISTANCE;
    }
  }

  up() {
    if (this.heading === DOWN && this.segments.length > 1) {
      this.collisionDetected = true;
    } else if (this.heading !== DOWN) {
      this.heading = UP;
    }
  }

  down() {
    if (this.heading === UP && this.segments.length > 1) {
      this.collisionDetected = true;
    } else if (this.heading !== UP) {
      this.heading = DOWN;
    }
  }

  left() {
    if (this.heading === RIGHT && this.segments.length > 1) {
      this.collisionDetected = true;
    } else if (this.heading !== RIGHT) {
      this.heading = LEFT;
    }
  }

  right() {
    if (this.heading === LEFT && this.segments.length > 1) {
      this.collisionDetected = true;
    } else if (this.heading !== LEFT) {
      this.heading = RIGHT;
    }
  }

  checkFoodCollision(food) {
    return (this.head.x === food.position[0] && this.head.y === food.position[1]);
  }

  checkWallCollision(screenWidth, screenHeight, topBarrier = 0) {
    return (this.head.x < 0 || 
            this.head.x >= screenWidth || 
            this.head.y < topBarrier ||  // Top barrier instead of 0
            this.head.y >= screenHeight);
  }

  checkSelfCollision() {
    if (this.collisionDetected) {
      return true;
    }

    const headPos = [this.head.x, this.head.y];
    for (let i = 1; i < this.segments.length; i++) {
      const segment = this.segments[i];
      if (headPos[0] === segment.x && headPos[1] === segment.y) {
        return true;
      }
    }
    return false;
  }

  reset() {
    this.segments = [];
    this.createSnake();
    this.head = this.segments[0];
    this.heading = RIGHT;
    this.collisionDetected = false;
  }

  draw(ctx) {
    ctx.fillStyle = '#22c55e'; // Green color
    this.segments.forEach(segment => {
      ctx.fillRect(segment.x, segment.y, this.cellSize - 1, this.cellSize - 1);
    });
  }
}